ProcessOptimizer/model_systems/color_pH.py

```diff
@@ -0,0 +1,339 @@
+
+    # If arbitury Keyword arg verbose=True then print out the below
+    if verbose == True:
+
+        decoration = '-'*20
+
+        print(decoration)
+        print('LAB Value Input')
+        print(decoration)
+        print(f"LAB1: {Lab1['L']}, {Lab1['a']}, {Lab1['b']}")
+        print(f"LAB2: {Lab2['L']}, {Lab2['a']}, {Lab2['b']}")
+        print(decoration)
+        print('Outputs')
+        print(decoration)
+        print(f'a1Prime: {round(a1Prime, 4)}')
+        print(f'a2Prime: {round(a2Prime, 4)}')
+        print(decoration)
+        print(f'cBar: {round(cBar, 4)}')
+        print(f'cBar7: {round(cBar7, 4)}')
+        print(decoration)
+        print(f'c1Prime: {round(c1Prime, 4)}')
+        print(f'c2Prime: {round(c2Prime, 4)}')
+        print(decoration)
+        print(f'h1Prime: {round(h1Prime, 4)}')
+        print(f'h2Prime: {round(h2Prime, 4)}')
+        print(decoration)
+        print(f'(abs)h1Prime - h2Prime: {round(abs(h1Prime - h2Prime), 4)}')
+        print(decoration)
+        print(f'hBarPrime: {round(hBarPrime, 4)}')
+        print(f'g: {round(g, 4)}')
+        print(f't: {round(t, 4)}')
+        print(f'sL: {round(sL, 4)}')
+        print(f'sC: {round(sC, 4)}')
+        print(f'sH: {round(sH, 4)}')
+        print(f'rT: {round(rT, 4)}')
+        print(decoration)
+        print(f'DE2000: {round(DE2000, 4)}')
+
+    if test == True:
+        a1Prime = round(a1Prime, 4)
+        a2Prime = round(a2Prime, 4)
+        c1Prime = round(c1Prime, 4)
+        c2Prime = round(c2Prime, 4)
+        h1Prime = round(h1Prime, 4)
+        h2Prime = round(h2Prime, 4)
+        hBarPrime = round(hBarPrime, 4)
+        g = round(g, 4)
+        t = round(t, 4)
+        sL = round(sL, 4)
+        sC = round(sC, 4)
+        sH = round(sH, 4)
+        rT = round(rT, 4)
+        DE2000 = round(DE2000, 4)
+
+        return a1Prime, a2Prime, c1Prime, c2Prime, h1Prime, h2Prime, hBarPrime, g, t, sL, sC, sH, rT, DE2000
+
+    else:
+        return DE2000
+
+
+def find_closest_match(file_name, target_vector, target_columns):
+    # make docstring for this function
+    '''
+    file_name: name of the csv file
+    target_vector: vector of target values
+    target_columns: list of column names for the target values
+    
+    returns: row with the closest match to the target vector
+    '''
+    data = []
+    with open(file_name, 'r') as f:
+        reader = csv.reader(f)
+        headers = next(reader)
+        for row in reader:
+            data.append(row)
+
+    # Convert data to numpy array for easier manipulation
+    data = np.array(data)
+
+    # Get the indices of the target columns
+    target_indices = [headers.index(col) for col in target_columns]
+
+    # Extract the data for the target columns
+    target_data = data[:, target_indices].astype(float)
+
+    # Calculate the absolute differences for column 'B'
+    differences_B = np.abs(target_data[:, 0] - target_vector[0])
```

Comment:
This assumes the difference in B is more important than the difference in C, is that on purpose?

We could try and find the difference in all dimensions, and then condense that into one number. That would also generalise to more than two parameters.

But we have to deal with normalisation to do that, which is probably more headaches than it is worth at the moment. So let's just be explicit about there being two inputs, and about one being more important.

Reply:
This is a very good point. And a sharp observation (!).
Yes, I deem "B" (acid-to-base-ratio) to be more important than the volume of indicator. You might call it "a helping hand to direct the optimization..."